### ISSUE TITLE
Required does not work with partial

### ISSUE THREAD

**Issue by q0w** [microsoft/pyright#3468]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
partial does not take in account Required and NotRequired.
**To Reproduce**
```python
from typing_extensions import Unpack, Required, NotRequired
from typing import TypedDict, overload, Callable, Optional
from functools import partial


class Kw(TypedDict):
    id: Required[str]
    std: Required[str]

class PrivateKw(Kw):
    private_key: Required[str]

class PublicKw(Kw):
    public_key: Required[str]


@overload
def add(**kwargs: Unpack[PublicKw]) -> None: ...

@overload
def add(**kwargs: Unpack[PrivateKw]) -> None: ...


def add(**kwargs):
    print(kwargs)


def req(f: Callable[..., None], key: Optional[str]) -> partial[None]:
    if key:
        return partial(f, private_key=key)
    return partial(f, public_key=key)


add(id="str", private_key='dasd', std='dasd')  # no errors, as expected
add(id="str", private_key='dasd')  # error: missing std kwarg, as expected
req(add, 'private-key')(id='str')  # no errors: missing std kwarg
```
**Expected behavior**
```python
req(add, 'private-key')(id='str')  #  should raise the error because of missing `std` argument
```

**VS Code extension or command-line**
command-line `v1.1.247`

**Comment by erictraut** [comment]
This behavior is expected.

`partial` is a special-cased when used in a direct call expression to transform the input argument (a function) into another function. You are using `partial` in a way where this special casing does not (and cannot) apply. You're using it as a return type annotation in `req`. Used in this way, it's simply a specialized class that doesn't encapsulate the input signature of a function. In this case, it falls back on the definition of the `__call__` method within the `partial` class:

```python
    def __call__(__self, *args: Any, **kwargs: Any) -> _T: ...
```

That explains why the call on the final line doesn't generate any errors.

**Comment by q0w** [comment]
Could I annotate in another way req function return type through Callable to make it work?

**Comment by erictraut** [comment]
If you want to support arbitrary function signatures, I can't think of any way to make it work with the current type system. If it were a positional parameter, you could use a `ParamSpec` and `Concatenate`, but those do not work with keyword parameters.

**Comment by q0w** [comment]
Thx.
